crypto/fipsmodule/sha/internal.h

```diff
@@ -92,6 +105,54 @@ void sha512_block_data_order(uint64_t *state, const uint8_t *in,
 #define KECCAK1600_ASM
 #endif
 
+// SHAx_Init_from_state is a low-level function that initializes |sha| with a
+// custom state. |h| is the hash state in big endian. |n| is the number of bits
+// processed at this point. It must be a multiple of |SHAy_CBLOCK*8|,
+// where SHAy=SHA1 if SHAx=SHA1, SHAy=SHA256 if SHAx=SHA224 or SHA256, and
+// SHAy=SHA512 otherwise.
+// This function returns one on success and zero on error.
+// This function is for internal use only and should never be directly called.
+OPENSSL_EXPORT int SHA1_Init_from_state(
+    SHA_CTX *sha, const uint8_t h[SHA1_CHAINING_LENGTH], uint64_t n);
+OPENSSL_EXPORT int SHA224_Init_from_state(
+    SHA256_CTX *sha, const uint8_t h[SHA224_CHAINING_LENGTH], uint64_t n);
+OPENSSL_EXPORT int SHA256_Init_from_state(
+    SHA256_CTX *sha, const uint8_t h[SHA256_CHAINING_LENGTH], uint64_t n);
+OPENSSL_EXPORT int SHA384_Init_from_state(
+    SHA512_CTX *sha, const uint8_t h[SHA384_CHAINING_LENGTH], uint64_t n);
+OPENSSL_EXPORT int SHA512_Init_from_state(
+    SHA512_CTX *sha, const uint8_t h[SHA512_CHAINING_LENGTH], uint64_t n);
+OPENSSL_EXPORT int SHA512_224_Init_from_state(
+    SHA512_CTX *sha, const uint8_t h[SHA512_224_CHAINING_LENGTH], uint64_t n);
+OPENSSL_EXPORT int SHA512_256_Init_from_state(
+    SHA512_CTX *sha, const uint8_t h[SHA512_256_CHAINING_LENGTH], uint64_t n);
+
+// SHAx_get_state is a low-level function that exports the hash state in big
+// endian into |out_n| and the number of bits processed at this point in
+// |out_n|. |SHAx_Final| must not have been called before (otherwise results
+// are not guaranteed). Furthermore, the number of bytes processed by
+// |SHAx_Update| must be a multiple of the block length |SHAy_CBLOCK| and
+// must be less than 2^61 (otherwise it fails). See comment above about
+// SHAx_Init_from_state for the definition of SHAy.
+// This function returns one on success and zero on error.
+// This function is for internal use only and should never be directly called.
+OPENSSL_EXPORT int SHA1_get_state(
```

Comment:
Does not need to be exported symbols(?)

Reply:
Unfortunately, we need to export them for the tests to work, e.g., https://github.com/aws/aws-lc/pull/1574/files#diff-a486879d22cdbac91336850b85c6f1c88b0725a1115b3d2b7674326925c17225R486-R490

Reply:
Ahh you generate test cases by macro. Fair.
My lazy ctrl+F search didn't catch that.